### PR TITLE
feat: build-only / install-skip-build modes for VHD-prebuilt GPU kernel module

### PR DESCRIPTION
**Pull request by ganeshkumarashok**
## What
Split `install.sh` so the expensive NVIDIA DKMS **kernel-module compile** can be hoisted to VHD build time, removing it from the node-provisioning critical path.

- `build_kernel_module` (device-independent: nouveau blacklist, overlayfs + `nvidia-installer --dkms`, lib staging, ld config) is separated from `device_init` (device-dependent: `nvidia-modprobe`/`nvidia-smi`, fabric manager, containerd runtime config, dev-char symlinks).
- New modes via `entrypoint.sh`:
  - `build-only` (`AKSGPU_BUILD_ONLY=1`): compile + DKMS-register + stage libs, **no device access** (safe on a GPU-less Packer builder), write `/opt/azure/aks-gpu/dkms-marker`.
  - `install-skip-build` (`AKSGPU_SKIP_KERNEL_BUILD=1`): skip recompile, run only device-init at node boot.
  - default `install`: unchanged legacy full path.
- **Marker re-validation** (kernel + driver_version + driver_kind) gates the skip-build fast path; any mismatch (kernel upgrade, or a shared VHD whose baked driver kind/version differs) falls back to a full build.
- `remove_stale_baked_driver` unloads loaded nvidia modules and removes a stale registered DKMS tree + relocated libs before a full rebuild, so a CUDA-baked VHD booting on a GRID node can't collide with the boot-time installer. No-op on today's VHDs.

## Why
The `nvidia-installer --dkms` compile is ~100s on the GPU provisioning critical path. Pre-baking it into the VHD removes it from every node boot.

## Consumer
AgentBaker uses these modes (build-only at VHD bake, install-skip-build at CSE): Azure/AgentBaker PR for `gpu-prebake-cuda-driver`.

## Safety
Default behaviour is unchanged when neither env var is set and no marker exists.

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -27,87 +64,208 @@ for apt_package in $NVIDIA_PACKAGES; do
...
+    umount -l "/usr/lib/${ARCH}-linux-gnu" || true
+    umount -l /tmp/overlay || true
+    rm -r /tmp/overlay || true
+    set -e
+
+    # set up overlayfs to change install location of nvidia libs from /usr/lib/$ARCH-linux-gnu to /usr/local/nvidia
+    # add an extra layer of indirection via tmpfs because it's not possible to have an overlayfs on an overlayfs (i.e., inside a container)
+    mkdir /tmp/overlay
+    mount -t tmpfs tmpfs /tmp/overlay
+    mkdir /tmp/overlay/{workdir,lib64}
```
Comment (timmy-wright) [inline install.sh]: **Trap misses partial overlay mount (Simon)**

If the `tmpfs` mount succeeds but the subsequent `overlay` mount fails, `OVERLAY_MOUNTED` is never set to `1` (that happens on the next line). The EXIT trap then skips `cleanup_overlay`, leaving `/tmp/overlay` mounted in the finished VHD.

Suggestion: set `OVERLAY_MOUNTED=1` immediately after the `tmpfs` mount and make `cleanup_overlay` tolerate a missing overlay layer (check with `mountpoint -q` before unmounting each layer individually).

File: install.sh [hunk 2]
```diff
@@ -4,21 +4,58 @@ set -euxo pipefail
...
+# Track overlay/tmpfs state so a build-time exit can never leave dangling mounts in the VHD.
+OVERLAY_MOUNTED=0
+cleanup_overlay() {
+    set +e
+    if [ "${OVERLAY_MOUNTED}" = "1" ]; then
+        umount -l "/usr/lib/${ARCH}-linux-gnu" || true
+        umount /tmp/overlay || true
+        rm -r /tmp/overlay || true
+        OVERLAY_MOUNTED=0
+    fi
```
Comment (timmy-wright) [inline install.sh:62]: **`cleanup_overlay` clears flag even when unmount fails (Book)**

`OVERLAY_MOUNTED=0` is set unconditionally inside the `if` block. If either `umount` call fails (e.g. mount is busy), the flag is cleared anyway, so the EXIT trap won't retry and a dangling mount can survive into the VHD.

Suggestion: only reset the flag after confirming the mountpoints are gone (e.g. `mountpoint -q` check, or inspect `/proc/self/mountinfo`).

File: install.sh [hunk 3]
```diff
@@ -27,87 +64,208 @@ for apt_package in $NVIDIA_PACKAGES; do
...
+    cp /opt/gpu/71-nvidia-char-dev.rules /lib/udev/rules.d/71-nvidia-dev-char.rules
+    /usr/bin/nvidia-ctk system create-dev-char-symlinks --create-all
+}
+
+write_dkms_marker() {
+    mkdir -p "$(dirname "${DKMS_MARKER_FILE}")"
+    cat > "${DKMS_MARKER_FILE}" <<EOF
+kernel=${KERNEL_NAME}
+driver_version=${DRIVER_VERSION}
+driver_kind=${DRIVER_KIND}
```
Comment (timmy-wright) [inline install.sh:186]: **build-only path not transactional around marker write (Book)**

If `build_kernel_module` succeeds but `write_dkms_marker` fails (disk full, permission error), the VHD ships with a prebaked DKMS/lib state but no marker. The boot-time path sees no marker, skips `remove_stale_baked_driver`, and the fresh `build_kernel_module` call collides with the stale prebaked DKMS state.

Suggestion: write the marker to a temp file first, then `mv` it atomically. On any marker-write failure, roll back the prebaked DKMS/lib state before exiting.

File: install.sh [hunk 4]
```diff
@@ -27,87 +64,208 @@ for apt_package in $NVIDIA_PACKAGES; do
...
+# backstop and will hard-fail if devices are genuinely in use.
+unload_nvidia_modules() {
+    local mod
+    for mod in nvidia_uvm nvidia_drm nvidia_modeset nvidia_peermem nvidia; do
+        if lsmod 2>/dev/null | grep -q "^${mod} "; then
+            echo "aks-gpu: unloading stale module ${mod}"
+            rmmod "${mod}" 2>/dev/null || modprobe -r "${mod}" 2>/dev/null || true
+        fi
+    done
+}
```
Comment (timmy-wright) [inline install.sh:216]: **Unload failure silently ignored on active node (Book)**

All `rmmod`/`modprobe -r` failures are swallowed with `|| true`. On an active GPU node these calls will likely fail due to open device handles — execution continues into `dkms remove` (also `|| true`), which can leave a partially-unloaded, partially-deregistered driver stack.

Suggestion: after the unload loop, assert all target modules are actually gone. If any remain, abort with a clear error rather than attempting DKMS removal against a live module.

File: install.sh [hunk 5]
```diff
@@ -27,87 +64,208 @@ for apt_package in $NVIDIA_PACKAGES; do
...
+# the legacy install path is unchanged.
+remove_stale_baked_driver() {
+    command -v dkms >/dev/null 2>&1 || return 0
+
+    # Versions of the nvidia DKMS module currently registered on the host, parsed from
+    # `dkms status` (handles both "nvidia/<ver>," and legacy "nvidia, <ver>," formats).
+    local registered_versions
+    registered_versions="$(dkms status 2>/dev/null \
+        | sed -n 's#^nvidia[,/][[:space:]]*\([^,]*\).*#\1#p' \
+        | sort -u || true)"
```
Comment (timmy-wright) [inline install.sh]: **`dkms status` parser captures trailing suffix (Simon)**

`dkms status` can emit lines like `nvidia/570.86.15: added` (in addition to the installed form). The current sed regex captures everything up to the first comma, so `nvidia/570.86.15: added` yields `570.86.15: added` (colon + suffix included). `dkms remove "nvidia/570.86.15: added"` is a silent no-op under `|| true`, leaving the stale DKMS registration.

Suggestion: stop capture at `:` or space too — e.g. `\([^,: ]*\)`.

File: install.sh [hunk 6]
```diff
@@ -27,87 +64,208 @@ for apt_package in $NVIDIA_PACKAGES; do
...
+
+    local v
+    for v in ${registered_versions}; do
+        echo "aks-gpu: removing stale baked nvidia DKMS module ${v} (node needs ${DRIVER_KIND} ${DRIVER_VERSION})"
+        dkms remove "nvidia/${v}" --all || true
+    done
+
+    # Drop the baked install's relocated libs + loader config + marker so only the driver we
+    # are about to build ends up on the path. build_kernel_module recreates the libs/conf.
+    rm -f /etc/ld.so.conf.d/nvidia.conf || true
```
Comment (timmy-wright) [inline install.sh]: **Early return skips marker/lib cleanup when DKMS state already absent (Simon + Book)**

If the marker exists but DKMS state is already partially missing (prior failed cleanup), `registered_versions` is empty and the function returns early — leaving the stale marker, `ld.so.conf.d/nvidia.conf`, and `${GPU_DEST}/lib64` on disk. The boot-time `build_kernel_module` then runs without having cleared those artefacts and may collide.

Suggestion: always remove marker/config/libs; gate only the `dkms remove` loop on non-empty `registered_versions`.

File: install.sh [hunk 7]
```diff
@@ -27,87 +64,208 @@ for apt_package in $NVIDIA_PACKAGES; do
...
-        NVIDIA_FM_ARCH="sbsa"
+install_nvidia_container_toolkit
+
+if [ "${AKSGPU_SKIP_KERNEL_BUILD}" = "1" ] && baked_marker_matches; then
+    # Node boot, prebuilt module valid for this kernel+driver: skip recompilation, ensure the
+    # baked module is loadable, then run the device-dependent steps only.
+    echo "aks-gpu: skip-kernel-build mode (using module prebuilt in VHD for kernel ${KERNEL_NAME})"
+    ldconfig
+    dkms status
+    modinfo -k "$KERNEL_NAME" nvidia
```
Comment (timmy-wright) [inline install.sh]: **Skip-build hard-fails on corrupt prebake instead of falling back (Simon + Book)**

`modinfo -k "$KERNEL_NAME" nvidia` only checks file metadata — not whether the module is actually loadable. If the baked module or lib state is corrupt/incomplete, `modinfo` (or the subsequent `dkms status`) exits nonzero and node provisioning dies under `set -e` with no recovery. Every node booted from a bad VHD bricks.

Suggestion: wrap the fast-path validation in an explicit check. On any failure, fall back: `remove_stale_baked_driver && build_kernel_module`. The fast path should be opportunistic, not a hard requirement.

**Review by timmy-wright — COMMENTED** [review]
## Squad Review — aks-gpu PR #162

**Reviewers:** Simon (defects) · Book (resilience) · Wash (testing) · Zoe (security: no findings)

The design is sound — separating `build_kernel_module` / `device_init` is the right shape. Two high-severity correctness bugs can brick nodes at scale; several medium-severity resilience gaps in the cleanup/stale-driver paths.

### 🔴 High

- **Trap misses partial overlay mount** (`install.sh` ~line 90, inline) — `OVERLAY_MOUNTED` not set when overlay mount fails after tmpfs succeeds; EXIT trap skips cleanup.
- **Skip-build hard-fails on corrupt prebake** (`install.sh` ~line 253, inline) — `modinfo` passes for corrupt modules; provisioning dies with no fallback rebuild.
- **Mode selection entirely untested** (Wash) — no shell tests for `build-only`/`install-skip-build`/unknown-action. Wrong env propagation or accidental full build at VHD bake time is undetectable. Suggest bats tests with mocked `nsenter`.

### 🟡 Medium

- **`cleanup_overlay` clears flag on unmount failure** (`install.sh` ~line 38, inline)
- **DKMS status parser captures trailing suffix** (`install.sh` ~line 209, inline) — `dkms remove` becomes no-op for `nvidia/ver: added` lines.
- **Early return skips marker/lib cleanup when DKMS state absent** (`install.sh` ~line 222, inline)
- **build-only not transactional around marker write** (`install.sh` ~line 152, inline)
- **Unload failure ignored on active node** (`install.sh` ~line 186, inline)
- **Marker format edge cases untested** (Wash) — blank `kernel=`, CRLF, duplicate keys cause silent fallback rebuilds with no test coverage.
- **Fallback path not CI-coverable without GPU** (Wash) — fully stubbable on a vanilla runner; `build_kernel_module`, `dkms`, `modinfo`, `ldconfig` can all be mocked.

### 🟢 Low

- **Observability thin for fast-path failures** (Wash) — log which marker field mismatched, what DKMS versions were removed.

**Comment by ganeshkumarashok** [comment]
Folded in the squad-review fixes + simplification (commit 8ee4482, install.sh −89/+48):
- Removed `remove_stale_baked_driver` + `unload_nvidia_modules` — `nvidia-installer -s` already uninstalls a prior runfile/DKMS driver before installing (kills the dkms-parser, unload-swallow, and early-return-cleanup findings).
- Opportunistic fast path (`fast_path_ok`) with fallback instead of `set -e` hard-fail — fixes the High "corrupt prebake bricks the node".
- `build_and_mark` always rewrites the marker (no perpetual-rebuild loop); mountpoint-driven overlay cleanup; atomic marker write.

Verify before merge: confirm `nvidia-installer -s` cleanly swaps a mismatched baked driver (GRID node on a CUDA-baked VHD) in e2e.

**Comment by xuexu6666** [comment]
### Code review

The new runtime logic has no automated guard: the repo has no unit tests and CI only runs `docker buildx build` (no execution of `install.sh`/`entrypoint.sh`, no shellcheck). So the mode dispatch (`build-only` / `install-skip-build` / `install`) and the marker write/parse/validate + fallback paths are verified only by a manual cross-repo AgentBaker e2e. It's pure shell branching and mockable on a GPU-less runner — a small bats harness (or even just a shellcheck CI step) would cover it.

**Review by awesomenix — APPROVED** [review]
